pkg/resolution/resolver/hub/resolver.go

```diff
@@ -126,36 +160,126 @@ func (r *Resolver) Resolve(ctx context.Context, params []pipelinev1beta1.Param)
 	// #nosec G107 -- URL cannot be constant in this case.
-	resp, err := http.Get(url)
+	resp, err := http.Get(apiEndpoint)
 	if err != nil {
-		return nil, fmt.Errorf("error requesting resource from hub: %w", err)
+		return fmt.Errorf("error requesting resource from Hub: %w", err)
 	}
 	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("requested resource '%s' not found on hub", url)
+		return fmt.Errorf("requested resource '%s' not found on hub", apiEndpoint)
 	}
 	defer func() {
 		_ = resp.Body.Close()
 	}()
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return nil, fmt.Errorf("error reading response body: %w", err)
+		return fmt.Errorf("error reading response body: %w", err)
 	}
-	hr := hubResponse{}
-	err = json.Unmarshal(body, &hr)
+
+	err = json.Unmarshal(body, v)
 	if err != nil {
-		return nil, fmt.Errorf("error unmarshalling json response: %w", err)
+		return fmt.Errorf("error unmarshalling json response: %w", err)
+	}
+
+	return nil
+}
+
+func resolveCatalogName(paramsMap, conf map[string]string) (string, error) {
+	var configTHCatalog, configAHTaskCatalog, configAHPipelineCatalog string
+	var ok bool
+
+	if configTHCatalog, ok = conf[ConfigTektonHubCatalog]; !ok {
+		return "", fmt.Errorf("default Tekton Hub catalog was not set during installation of the hub resolver")
+	}
+	if configAHTaskCatalog, ok = conf[ConfigArtifactHubTaskCatalog]; !ok {
+		return "", fmt.Errorf("default Artifact Hub task catalog was not set during installation of the hub resolver")
+	}
+	if configAHPipelineCatalog, ok = conf[ConfigArtifactHubPipelineCatalog]; !ok {
+		return "", fmt.Errorf("default Artifact Hub pipeline catalog was not set during installation of the hub resolver")
 	}
-	return &ResolvedHubResource{
-		Content: []byte(hr.Data.YAML),
-	}, nil
+	if _, ok := paramsMap[ParamCatalog]; !ok {
+		switch paramsMap[ParamType] {
+		case ArtifactHubType:
+			switch paramsMap[ParamKind] {
+			case "task":
+				return configAHTaskCatalog, nil
+			case "pipeline":
+				return configAHPipelineCatalog, nil
+			default:
+				return "", fmt.Errorf("failed to resolve catalog name with kind: %s", paramsMap[ParamKind])
+			}
+		case TektonHubType:
+			return configTHCatalog, nil
+		default:
+			return "", fmt.Errorf("failed to resolve catalog name with type: %s", paramsMap[ParamType])
+		}
+	}
+
+	return paramsMap[ParamCatalog], nil
+}
+
+// the Artifact Hub follows the semVer (i.e. <major-version>.<minor-version>.0)
+// the Tekton Hub follows the simplified semVer (i.e. <major-version>.<minor-version>)
+// for resolution request with "artifact" type, we append ".0" suffix if the input version is simplified semVer
+// for resolution request with "tekton" type, we only use <major-version>.<minor-version> part of the input if it is semVer
+func resolveVersion(version, hubType string) (string, error) {
+	semVer := strings.Split(version, ".")
+	resVer := version
+
+	if hubType == ArtifactHubType && len(semVer) == 2 {
+		resVer = version + ".0"
```

Comment:
I'm a bit concerned about doing this - i.e. if someone takes this version and tries to fetch the config, will this guarantee to always work? i.e. if artifacthub publishes `0.1`, will I *always* be able to fetch `0.1.0` and will this *always* map to `0.1`? Or could a different `0.1.0` be added after?

There's an argument that maybe the resolver shouldn't try to interpret semver and just always handle versions as-is.

Reply:
Hi @wlynch, thanks for your input and this is a good question 😀 . The short answer is **yes**. If artifacthub publishes 0.1, and I will  always be able to fetch 0.1.0

The artifact hub will always convert the [resource yaml version](https://github.com/tektoncd/catalog/blob/962f9005350b1acb359558b6c5aafc420e0beb46/task/git-clone/0.8/git-clone.yaml#L6) to semVer (i.e. `0.6` -> `0.6.0`) and the Hub uses the semVer to index the corresponding resource.
 
So technically you can publish a resource with yaml version `0.6` first (which is converted to `0.6.0` in the Hub), and then publish another yaml version of `0.6.0`, it just overrides the previous `0.6.0` version in the Hub. But the API will always return the latest `0.6.0`

The main reason I put such conversion is for backward compatibility, so that current users can keep version 0.6 in the resolver, and the type is default to `artifact`

Reply:
I'd lean towards dropping this for the new artifacthub impl:

1. semver with a missing minor/patch version generally uses the latest version matching the prefix (i.e. the golang 1.19 docker image refers to the latest patch version 1.19.2) - I'm not sure we want to have the resolver implement this kind of lookup behavior.
2. I'm worried about version skew for ArtifactHub URLs - e.g. https://artifacthub.io/packages/tekton-task/tekton-catalog-tasks/git-clone/0.8.0 (even though this points to the 0.8 git tag), but https://artifacthub.io/packages/tekton-task/tekton-catalog-tasks/git-clone/0.8 404s.

It'll probably be simpler long term for the resolver to interpret the user provided version tag as-is, so it always matches the upstream state (up to you if you want to preserve the old behavior for compatibility).

Reply:
For 2, I don't think there is version skew for the Artifact Hub URL since all the valid URLs will use semVer no matter how do you tag the repo, and URLs like https://artifacthub.io/packages/tekton-task/tekton-catalog-tasks/git-clone/0.8 will always return 404s.

Reply:
Thoughts? @jerop @abayer @vdemeester?